### ISSUE TITLE
ask for pivot feature

### ISSUE THREAD

**Issue by l1t1** [ClickHouse/ClickHouse#50477]
please add the pivot/unpivot statement in clickhouse. 
### Discussed in https://github.com/ClickHouse/ClickHouse/discussions/50473

<div type='discussions-op-text'>

<sup>Originally posted by **l1t1** June  2, 2023</sup>
oracle, duckdb
https://docs.oracle.com/cd/E11882_01/server.112/e41084/statements_10002.htm#SQLRF01702
https://duckdb.org/docs/sql/statements/pivot</div>

**Comment by l1t1** [comment]
duckdb design
https://github.com/duckdb/duckdb/pull/6961#issuecomment-1554744326

**Comment by danthegoodman1** [comment]
Bumping as this is super useful and painful without

**Comment by alanpaulkwan** [comment]
+1 +1 +1

**Comment by anjmao** [comment]
Aggregate combinators seems to solve pivot https://clickhouse.com/blog/aggregate-functions-combinators-in-clickhouse-for-arrays-maps-and-states

```sql
SELECT
    countIf((status = 'confirmed') AND (confirm_time > (create_time + toIntervalMinute(1)))) AS num_confirmed_checked,
    sumIf(total_amount, (status = 'confirmed') AND (confirm_time > (create_time + toIntervalMinute(1)))) AS confirmed_checked_amount,
    countIf(status = 'declined') AS num_declined,
    sumIf(total_amount, status = 'declined') AS dec_amount,
    avgIf(total_amount, status = 'declined') AS dec_average
FROM payments

┌─num_confirmed_checked─┬─confirmed_checked_amount─┬─num_declined─┬────────dec_amount─┬───────dec_average─┐
│                    39 │        11195.98991394043 │           50 │ 10780.18000793457 │ 215.6036001586914 │
└───────────────────────┴──────────────────────────┴──────────────┴───────────────────┴───────────────────┘
```

**Comment by alanpaulkwan** [comment]
@anjmao doesn't this require manually enumerating values? What if I had 50 distinct values

**Comment by rnesytov** [comment]
I've tried solution with argMaxIf (get last values by created_at for specific column value) with 200+ values/columns. Got an issue with memory usage, every new value/column requires more and more memory. 
So, one could

**Comment by HongYangze** [comment]
In the scenario of converting narrow tables to wide tables, which is a very common scenario for column-oriented databases , PIVOT is a very important and common feature. Since we often don't know how many columns there are in advance, or there are too many columns, it would be very troublesome to use the current alternative methods (map or Aggregate combinators). The PIVOT method similar to duckdb is very concise and easy to use. Sincerely hope that developers can add this feature.

**Comment by fb64** [comment]
BigQuery also [supports PIVOT / UNPIVOT](https://cloud.google.com/bigquery/docs/reference/standard-sql/query-syntax#pivot_operator) operator

**Comment by alexey-milovidov** [comment]
`anyMap` does it: https://pastila.nl/?00950c9a/60216367f77a4479e48df72a51bc908d

**Comment by fb64** [comment]
Indeed, `anyMap` and `arrayJoin` can help with Pivot / Unpivot tasks.
But dedicated functions would be more explicit, user-friendly and easier to work with (IMHO).

**Comment by danthegoodman1** [comment]
@alexey-milovidov anyMap doesn’t seem to be in the docs? Google results are sparse too

**Comment by alexey-milovidov** [comment]
This is because the `-Map` is an aggregate function combinator, similar to `-If`, `-Array`, etc: https://clickhouse.com/docs/en/sql-reference/aggregate-functions/combinators

**Comment by danthegoodman1** [comment]
I understand now. Unfortunately this requires a very deep knowledge of clickhouse that anyone evaluating against another solution will likely not stumble across (or have the knowledge).

**Comment by chriscomeau79** [comment]
This would be a good topic for a blog post or video.

**Comment by mneedham** [comment]
@chriscomeau79 I wrote up a blog post going through each of DuckDB's examples and showing how to do the equivalent in ClickHouse.

https://www.markhneedham.com/blog/2024/11/15/pivot-clickhouse-duckdb/
